basetest.pm

```diff
@@ -640,25 +640,29 @@ sub parse_serial_output_qemu {
     my $json = myjsonrpc::read_json($autotest::isotovideo);
 
     my $die = 0;
+    my %regexp_matched;
     # loop line by line
     for my $line (split(/^/, $json->{serial})) {
         chomp $line;
-        # only two keys allowed
-        for my $type (qw(soft hard)) {
-            for my $regexp (@{$failures->{$type}}) {
-                # If you want to match a simple string please be sure that you create it with quotemeta
-                if ($line =~ /$regexp/) {
-                    my $fail_type = 'softfail';
-                    if ($type eq 'hard') {
-                        $die       = 1;
-                        $fail_type = 'fail';
-                    }
-
-                    $self->record_testresult($fail_type);
-                    $self->record_resultfile('Serial Failure', "Serial error: $line", result => $fail_type);
-                    $self->{result} = $fail_type;
-
+        for my $regexp_table (@{$failures}) {
+            my $regexp  = $regexp_table->{pattern};
+            my $message = $regexp_table->{message};
+            my $type    = $regexp_table->{type};
+            die "Wrong type defined for serial failure. Only soft or hard allowed." if $type !~ /soft|hard/;
+            die "Not defined message key. Please, define one for each regular expression" if !defined $message;
+
+            # If you want to match a simple string please be sure that you create it with quotemeta
+            if (!exists $regexp_matched{$regexp} and $line =~ /$regexp/) {
```

Comment:
If you want to avoid consecutive regular exceptions then you need to save the exception position as well.

Reply:
Hmm, good point. Question is if we want to see same soft-failures even in case of multiple matches or just once and then it's possible to check serial output in case more details are required.